src/reader.rs

```diff
@@ -156,115 +160,114 @@ impl<R: Read> Block<R> {
         self.message_count -= 1;
         Ok(Some(item))
     }
 }
 
-/// Main interface for reading Avro formatted values.
-///
-/// To be used as an iterator:
-///
-/// ```no_run
-/// # use avro_rs::Reader;
-/// # use std::io::Cursor;
-/// # let input = Cursor::new(Vec::<u8>::new());
-/// for value in Reader::new(input).unwrap() {
-///     match value {
-///         Ok(v) => println!("{:?}", v),
-///         Err(e) => println!("Error: {}", e),
-///     };
-/// }
-/// ```
 pub struct Reader<'a, R> {
     block: Block<R>,
-    reader_schema: Option<&'a Schema>,
-    errored: bool,
+    reader_schema: Option<Cow<'a, Schema>>,
     should_resolve_schema: bool,
 }
 
-impl<'a, R: Read> Reader<'a, R> {
-    /// Creates a `Reader` given something implementing the `io::Read` trait to read from.
+impl<R: AsyncRead + Unpin + Send + 'static> Reader<'static, R> {
+    /// Creates a `Reader` given a reader `Schema` and something implementing the `tokio::io::AsyncRead` trait
+    /// to read from. Takes ownership of both objects.
+    ///
+    /// **NOTE** The avro header is going to be read automatically upon creation of the `Reader`.
+    pub async fn with_schema_owned(schema: Schema, reader: R) -> Result<Reader<'static, R>, Error> {
+        Self::with_schema_inner(Cow::Owned(schema), reader).await
+    }
+}
+
+impl<'a, R: AsyncRead + Unpin + Send + 'a> Reader<'a, R> {
+    /// Creates a `Reader` given something implementing the `tokio::io::AsyncRead` trait to read from.
     /// No reader `Schema` will be set.
     ///
     /// **NOTE** The avro header is going to be read automatically upon creation of the `Reader`.
-    pub fn new(reader: R) -> Result<Reader<'a, R>, Error> {
-        let block = Block::new(reader)?;
+    pub async fn new(reader: R) -> Result<Reader<'a, R>, Error> {
+        let block = Block::new(reader).await?;
         let reader = Reader {
             block,
             reader_schema: None,
-            errored: false,
             should_resolve_schema: false,
         };
         Ok(reader)
     }
 
-    /// Creates a `Reader` given a reader `Schema` and something implementing the `io::Read` trait
-    /// to read from.
-    ///
-    /// **NOTE** The avro header is going to be read automatically upon creation of the `Reader`.
-    pub fn with_schema(schema: &'a Schema, reader: R) -> Result<Reader<'a, R>, Error> {
-        let block = Block::new(reader)?;
+    async fn with_schema_inner(schema: Cow<'a, Schema>, reader: R) -> Result<Reader<'a, R>, Error> {
+        let block = Block::new(reader).await?;
         let mut reader = Reader {
             block,
             reader_schema: Some(schema),
-            errored: false,
             should_resolve_schema: false,
         };
         // Check if the reader and writer schemas disagree.
-        reader.should_resolve_schema = reader.writer_schema() != schema;
+        reader.should_resolve_schema = reader.writer_schema() != reader.reader_schema.as_ref().unwrap().borrow();
```

Comment:
does this need the unwrap? Like probably better to move this check to before the schema is embedded in the schema?

Reply:
Indeed I think the whole function simplifies to:

```rust
    async fn with_schema_inner(schema: Cow<'a, Schema>, reader: R) -> Result<Reader<'a, R>, Error> {
        let block = Block::new(reader).await?;
        Ok(Reader {
            should_resolve_schema: &block.writer_schema != schema.borrow(),
            block,
            reader_schema: Some(schema),
        })
    }
```